File: obstacle_avoidance/nodes/sonar_to_pointcloud_node.cpp
#include <ros/ros.h>
#include <sensor_msgs/PointCloud2.h>
#include <marine_acoustic_msgs/ProjectedSonarImage.h>
#include <pcl/point_cloud.h>
#include <pcl_conversions/pcl_conversions.h>

ros::Publisher point_cloud_pub;

void sonarCallback(const marine_acoustic_msgs::ProjectedSonarImage::ConstPtr& msg)
{
    pcl::PointCloud<pcl::PointXYZ> cloud;

    // Convert beam_directions and ranges to point cloud
    for(size_t i = 0; i < msg->beam_directions.size(); i++)
    {
        pcl::PointXYZ point;
        point.x = msg->ranges[i] * msg->beam_directions[i].x;
        point.y = msg->ranges[i] * msg->beam_directions[i].y;
        point.z = msg->ranges[i] * msg->beam_directions[i].z;
        cloud.push_back(point);
    }

    // Convert PCL point cloud to PointCloud2
    sensor_msgs::PointCloud2 output;
    pcl::toROSMsg(cloud, output);
    output.header.frame_id = "sonar_frame";
    output.header.stamp = ros::Time::now();

    point_cloud_pub.publish(output);
    ROS_INFO("Processing new sonar data...");
}

int main(int argc, char **argv)
{
    ros::init(argc, argv, "sonar_to_pointcloud_node");
    ros::NodeHandle nh;

    ros::Subscriber sonar_sub = nh.subscribe<marine_acoustic_msgs::ProjectedSonarImage>("/bluerov/oculus_m1200d/sonar_image_raw", 10, sonarCallback);
    point_cloud_pub = nh.advertise<sensor_msgs::PointCloud2>("/sonar_pointcloud", 10);

    ros::spin();

    return 0;
}

File: obstacle_avoidance/nodes/sonar_to_pointcloud_node.py

#!/usr/bin/env python

import rospy
from sensor_msgs.msg import PointCloud2, PointField
import sensor_msgs.point_cloud2 as pc2
import re
import numpy as np

def parse_sonar_data(filename):
    with open(filename, 'r') as f:
        content = f.read()

    # Parsing functions similar to our previous implementations
    # ... (for brevity, not showing here, but will be included in the script)

    # Extract beam directions, ranges, etc.
    # ... 

    return beam_directions, ranges

def numpy_to_pointcloud2(points, frame_id="base_link"):
    fields = [PointField('x', 0, PointField.FLOAT32, 1),
              PointField('y', 4, PointField.FLOAT32, 1),
              PointField('z', 8, PointField.FLOAT32, 1)]

    header = rospy.Header()
    header.stamp = rospy.Time.now()
    header.frame_id = frame_id

    return pc2.create_cloud(header, fields, points)

def sonar_data_to_pointcloud(filename):
    beam_directions, ranges = parse_sonar_data(filename)
    points = [r * np.array(direction) for direction, r in zip(beam_directions, ranges)]
    return points

if __name__ == '__main__':
    rospy.init_node('sonar_pointcloud_publisher')

    # Parse the sonar data from the provided file
    points = sonar_data_to_pointcloud("/path/to/your/projectedsonarimage.txt")

    pub = rospy.Publisher('/sonar_pointcloud', PointCloud2, queue_size=10)
    rate = rospy.Rate(1)  # 1 Hz

    while not rospy.is_shutdown():
        pointcloud_msg = numpy_to_pointcloud2(points)
        pub.publish(pointcloud_msg)
        rate.sleep()
